implementation of an automatic gpr file with

Fabien-Chouteau opened the pull request:
When adding or removing dependency with `alr with`, a list of `with` statement
for each project files of the dependencies can be automatically added to the
root project file:
```
-- begin auto-gpr-with --
--  This section was automatically added by Alire
with "libhello.gpr";
-- end auto-gpr-with --
project Test is
...
```

This feature can be permanently enabled or disabled with a local or global
configuration option:
```bash
alr config --global --set auto-gpr-with false
```

Crates with project files not compatible with this feature can disable it using
the `auto-gpr-with` entry:
```toml
auto-gpr-with=false
```

ref #313

Fabien-Chouteau committed:
First prototype implementation of an automatic gpr file with

Fabien-Chouteau committed:
Add user query and config option for Auto-gpr-with

Fabien-Chouteau committed:
Introduce Boolean properties

Fabien-Chouteau committed:
Alire.Properties.Labeled: check TOML_Kind for labels

Fabien-Chouteau committed:
auto-gpr-with: add crate entry to disable auto-gpr-with feature

Fabien-Chouteau committed:
testsuite: allow custom indexes to reference fixtures/crates/

This removes duplication of several sources and archives.

Fabien-Chouteau committed:
Bump index version to 0.3 for auto-gpr-with feature

Fabien-Chouteau committed:
auto-gpr-with: do not add default project file for extern crates

Fabien-Chouteau committed:
Bump alire-index in CI scripts

Fabien-Chouteau committed:
tests/workflows/auto-gpr-with: rename test crate to avoid conflicts with scripts

mosteo commented on doc/catalog-format-spec.md:
```diff
@@ -350,6 +350,13 @@ entries:
 
+ - `auto-gpr-with`: Boolean value that specifies if the project files of a
+   crate can be automatically 'withed' by the root project file (default is
+   true). This feature is meant to simply the process of using dependencies in
```
s/simply/simplify/

mosteo commented on src/alire/alire-properties-from_toml.ads:
```diff
@@ -26,7 +27,8 @@ package Alire.Properties.From_TOML with Preelaborate is
                           Website,
-                          Tags);
+                          Tags,
+                          Auto_GPR_With);
```
No need to change but for future occasions: my OCD is appeased by having these enums in alphabetical order if there's no reason for other ordering :-)

mosteo commented on .github/workflows/ci-linux.yml:
```diff
@@ -9,7 +9,7 @@ on:
 
 env:
-  alire_index: ""
+  alire_index: "git+https://github.com/alire-project/alire-index@devel-0.3"
```
This should not be necessary because the index is determined by `Alire.Index.Community_Branch`, so this will result in extra maintenance for no gain. In fact, it's a remainder of when we could not select an index in any other way. 

So, is it giving you some trouble that I'm missing?

Fabien-Chouteau replied:
Actually I made several attempts to get the testsuite to run, so this might not be necessary in the end.

mosteo commented on testsuite/drivers/helpers.py:
```diff
@@ -5,21 +5,6 @@
 import os
 
 
-# Check a file contains a concrete line
-def check_line_in(filename, line):
-    """
-    Assert that the `filename` text file contains at least one line that
-    contains `line`.
-    """
-    with open(filename, 'r') as f:
-        for l in f:
-            if l.rstrip() == line:
-                break
-        else:
-            assert False, 'Could not find {} in {}'.format(
-                repr(line), filename)
```
Don't we want this one just in case for the future? Or is there something wrong with it?

Fabien-Chouteau replied:
Ah yes, this should be in a separate commit, but this function is actually a duplicate. There is exactly the same one defined in the same file.

mosteo commented:
Looks great overall, Fabien, thanks. I just have a couple of doubts that I guess I'm missing by being out of context.

Fabien-Chouteau commented:
> Looks great overall, Fabien, thanks.

Thank you, it's a nice to have feature.

> I just have a couple of doubts that I guess I'm missing by being out of context.

Are you talking about the comments in your review, or do you have other doubts?

Fabien-Chouteau committed:
Review fixes

mosteo commented:
> Are you talking about the comments in your review, or do you have other doubts?

Just the comments you already answered :+1:

Fabien-Chouteau committed:
tests/get/external-tool-dependency: make the test more robust

On Windows when using msys2, the make seen by alr can be different than
the one seen from the testsuite scripts. The means that the current way
to identify make's version will fail in some cases.

This patch introduces a different way to workaround the different
version numbers of make.
